| ``Numeric Tower`` | ``Description`` |
| --------------|----------|
| Complex | It has real and imag parts. It has a function which computes the conjugate. All other methods except  `-`, `!-` are abstract (if a method is abstract then it must be defined in its subclass) |
| x//y | Floored quotient of x and y (results are rounded towards -inf) |
| int.bit_length() | number of bits necessary to represent an integer in binary, excluding sign and leading zeroes. |
| bit_count() | number of ones in binary representation of absolute value of integer |
| to_bytes() | returns array of bytes representing an integer |
| as_integer_ratio() | Ex. 10.as_integer_ratio()=(10,1) |

In [ ]:
num = -10
float = 0.8
print(num.as_integer_ratio())
print(float.as_integer_ratio())
nim=10
print(nim.bit_length())
print(nim.bit_count()) # does not work for older version

(-10, 1)
(3602879701896397, 4503599627370496)
4


AttributeError: 'int' object has no attribute 'bit_count'

| ``Expression`` | ``Result`` | ``Why?`` |
| --------------|----------| ---- |
| 0.8.as_integer_ratio() | (3602879701896397, 4503599627370496) | closest binary approximation |
| Fraction(0.8) | same as above | It is float under the hood |
| Fraction('0.8') | (4,5) | Calls decimal.Decimal('0.8'), then converts decimal to exact fraction |


| ``Feature`` | ``float`` | ``decimal.Decimal`` |
| ------- | ----- | --------------- |
| Storage base | Binary (base-2) | Decimal (base-10) |
| Precision | Approximate | Exact |
| Speed | Fast (hardware-supported) | Slower |
| Default availability | Built in | Must import (decimal) | 
| Suitable for | Science | Finance |